src/opt/horizontal_fusion_impl.cpp

```diff
@@ -0,0 +1,839 @@
+        char* dst           = input.get();
+        while(copy_bytes < total_bytes)
+        {
+            unsigned ins_ndx = 0;
+            for(auto&& ins : instrs)
+            {
+                unsigned long long bytes = bytes_per_slice[ins_ndx];
+                const char* src          = &(ins->get_literal().data()[slice_ndx * bytes]);
+                std::copy(src, src + bytes, dst);
+                dst += bytes;
+                copy_bytes += bytes;
+                ins_ndx++;
+            }
+            slice_ndx++;
+        }
+        shape new_shape{s.type(), new_lens};
+        auto new_literal = p_program->add_literal(literal(new_shape, input.get()));
+        instruction::replace_argument(output, ins0, new_literal, false);
+    }
+    else
+    {
+        ins0->set_shape({ins0->get_shape().type(), new_lens});
+        if(ins0->name() == "broadcast")
+        {
+            // workaround for a bad practice: broadcast has a broadcast_shape field.
+            uint64_t ax  = (any_cast<op::broadcast>(ins0->get_operator())).axis;
+            operation op = op::broadcast{ax, ins0->get_shape().lens()};
+            std::vector<shape> input_shapes;
+            input_shapes.push_back(ins0->inputs().at(0)->get_shape());
+            shape new_s = op.compute_shape(input_shapes);
+            ins0->set_operator(op);
+            ins0->set_shape(new_s);
+        }
+    }
+
+    if(output == base)
+        output->set_shape({output->get_shape().type(), base_lens});
+}
+
+// If ins and input only diff in one axis, return that axis.
+int horizontal_fusion_impl::find_unique_axis(instruction_ref ins, instruction_ref input)
+{
+    auto lens1 = ins->get_shape().lens();
+    auto lens2 = input->get_shape().lens();
+    if(lens1.size() != lens2.size())
+        return -1;
+    int count = 0;
+    int ndx   = 0;
+    int ret   = -1;
+    for(auto&& size : lens1)
+    {
+        if(size != lens2.at(ndx))
+        {
+            count++;
+            ret = ndx;
+        }
+        ndx++;
+    }
+    return (count == 1) ? ret : -1;
+}
+
+// find concat axis for ins.
+int horizontal_fusion_impl::find_axis(instruction_ref ins,
+                                      std::unordered_map<instruction_ref, bool>& is_common)
+{
+    int axis = -1;
+    for(auto&& input : ins->inputs())
+    {
+        if(is_common.find(input) != is_common.end())
+        {
+            int cur_axis = find_unique_axis(ins, input);
+            if((cur_axis == -1) || ((axis != -1) && (cur_axis != axis)))
+                return -1;
+            axis = cur_axis;
+        }
+    }
+    if(is_conv(ins) && (axis != get_channel_axis()))
+        return -1;
+
+    return axis;
+}
+
+// remove instructions in given vector except the first one.
+void horizontal_fusion_impl::remove_redundant_roots(std::vector<instruction_ref>& base_instrs)
+{
+    instruction_ref root_ins = base_instrs.at(0);
+    for(unsigned long ndx = 1; ndx < base_instrs.size(); ndx++)
```

Comment:
This can be written with for-range loop:

```cpp
for(auto base:range(base_instrs.begin()+1,base_instrs.end()))
{
...
}
```

Reply:
I don't understand.  Is there a range class in c++?

Reply:
This is in the `migraphx/ranges.hpp` header.

Reply:
fixed.